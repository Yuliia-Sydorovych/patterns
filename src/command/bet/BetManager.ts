import { IPlayer } from "../../observer/interface/IPlayer";

export class BetManager
{
  private player: IPlayer<number>;

  constructor(player: IPlayer<number>)
  {
    this.player = player;
  }

  public placeBet(amount: number): boolean
  {
    return this.player.placeBet(amount);
  }

  public updateBalance(payout: number): void
  {
    this.player.updateBalance(payout);
  }

  public checkBet(value: number): boolean
  {
    return this.player.checkBet(value);
  }
}
