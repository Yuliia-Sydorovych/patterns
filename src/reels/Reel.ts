export class Reel
{
  private symbols: string[];
  private currentSymbol: string;
  private spinning: boolean;
  private reelElement: HTMLElement | null;

  constructor(symbols: string[], reelIndex: number)
  {
    this.symbols = symbols;
    this.currentSymbol = symbols[0];
    this.spinning = false;
    this.reelElement = document.getElementById(`reel-${reelIndex}`);
  }

  public spin(): void
  {
    this.spinning = true;
    let index = 0;

    const interval = setInterval(() =>
    {
      index = Math.floor(Math.random() * this.symbols.length);
      
      this.currentSymbol = this.symbols[index];

      if (this.reelElement)
      {
        this.reelElement.textContent = this.currentSymbol;
      }

      if (!this.spinning)
      {
        clearInterval(interval);
      }
    }, 100);
  }

  public stop(): void
  {
    this.spinning = false;
  }

  public getCurrentSymbol(): string
  {
    return this.currentSymbol;
  }

  public isSpinning(): boolean
  {
    return this.spinning;
  }
}
