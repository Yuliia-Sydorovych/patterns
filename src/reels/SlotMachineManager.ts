import { ISlotMachine } from "./interface/ISlotMachine";

export class SlotMachineManager
{
    private slotMachine: ISlotMachine;
  
    constructor(slotMachine: ISlotMachine)
    {
      this.slotMachine = slotMachine;
    }
  
    public spin(onComplete: () => void): void
    {
      this.slotMachine.spinReels();
  
      const spinCheckInterval = setInterval(() =>
      {
        if (!this.slotMachine.areReelsSpinning())
        {
          clearInterval(spinCheckInterval);
          onComplete();
        }
      }, 100);
    }
  
    public calculatePayout(betAmount: number): number
    {
      return this.slotMachine.calculatePayout(betAmount);
    }
  }