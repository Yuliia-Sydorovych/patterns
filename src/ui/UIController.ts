import { IMediator } from "../mediator/interface/IMediator";
import { IUIController } from "./interface/IUIController";

export class UIController implements IUIController
{
    private mediator!: IMediator;
    private spinButton: HTMLButtonElement;
    private addFundsButton: HTMLButtonElement;
    private betButtons: NodeListOf<HTMLButtonElement>;
    private balanceElement: HTMLElement;
    private resultMessage: HTMLElement;
    private betAmount: number = 1;
  
    constructor()
    {
        this.spinButton = document.getElementById("spin-button") as HTMLButtonElement;
        this.addFundsButton = document.getElementById("add-funds-button") as HTMLButtonElement;
        this.betButtons = document.querySelectorAll(".bet-button") as NodeListOf<HTMLButtonElement>;
        this.balanceElement = document.getElementById("balance")!;
        this.resultMessage = document.getElementById("game-result")!;

        this.setupSpinButton();
        this.setupBetButtons();
        this.setupFundsButton();
    }

    public bindMediator(mediator: IMediator): void
    {
        this.mediator = mediator;
    }

    public updateGameResult(message: string): void
    {
        this.resultMessage.textContent = message;
    }

    public setSpinButtonState(enabled: boolean): void
    {
        this.spinButton.disabled = !enabled;
    }

    public setBetButtonsState(enabled: boolean): void
    {
        this.betButtons.forEach((button) =>
        {
            button.disabled = !enabled;
        });
    }

    public setBalanceValue(value: number): void
    {
        this.balanceElement.textContent = `Balance: $${value}`;
    }

    private setupSpinButton(): void
    {
        this.spinButton.addEventListener("click", () =>
        {
            this.mediator?.notify(this.spinButton, "spin", this.betAmount);

            this.setBetButtonsState(false);
            this.setSpinButtonState(false);
        });
    }

    private setupBetButtons(): void
    {
        this.betButtons.forEach((button) =>
        {
            button.addEventListener("click", () =>
            {
                this.selectBet(button);
            });
        });

        this.selectBet(this.betButtons[0]);
    }

    private selectBet(button: HTMLButtonElement): void
    {
        this.betButtons.forEach((btn) => btn.classList.remove("selected"));
        button.classList.add("selected");

        this.betAmount = parseInt(button.textContent!.replace('$', ''));
        (document.getElementById("bet-amount") as HTMLInputElement).value = this.betAmount.toString();

        this.mediator?.notify(button, "bet", this.betAmount);
    }

    private setupFundsButton(): void
    {
        this.addFundsButton.addEventListener("click", () =>
        {
            this.mediator?.notify(this.addFundsButton, "funds", 100);

            this.updateGameResult("You added $100. Spin to play!");
        });
    }
}
  