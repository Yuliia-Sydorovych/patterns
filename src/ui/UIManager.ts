import { IUIController } from "./interface/IUIController";

export class UIManager
{
    private uiController: IUIController;
  
    constructor(uiController: IUIController)
    {
      this.uiController = uiController;
    }
  
    public updateOnSpinStart(): void
    {
      this.uiController.setSpinButtonState(false);
      this.uiController.setBetButtonsState(false);
      this.uiController.updateGameResult("Spinning...");
    }
  
    public updateOnSpinEnd(payout: number): void
    {
      this.uiController.setSpinButtonState(true);
      this.uiController.setBetButtonsState(true);
  
      const message = payout > 0 ? `You won $${payout}!` : "You lost this round.";
      this.uiController.updateGameResult(message);
    }
  
    public showInsufficientFunds(): void
    {
      this.uiController.updateGameResult("Insufficient funds to place the bet.");
    }

    public updateGameResult(value: string): void
    {
      this.uiController.updateGameResult(value);
    }
  }