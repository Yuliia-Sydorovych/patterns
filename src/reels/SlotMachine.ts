import { ISlotMachine } from "./interface/ISlotMachine";
import { Reel } from "./Reel";

export class SlotMachine implements ISlotMachine
{
  private reels: Reel[];

  constructor(reels: Reel[])
  {
    this.reels = reels;
  }

  public spinReels(): void
  {
    this.reels.forEach((reel) => reel.spin());

    setTimeout(() =>
    {
      this.reels.forEach((reel) => reel.stop());
    }, 2001);
  }

  public areReelsSpinning(): boolean
  {
    return this.reels.some((reel) => reel.isSpinning());
  }

  public getFinalSymbols(): string[]
  {
    return this.reels.map((reel) => reel.getCurrentSymbol());
  }

  public calculatePayout(betAmount: number): number
  {
    const finalSymbols = this.getFinalSymbols();

    if (finalSymbols.every((symbol) => symbol === finalSymbols[0]))
    {
      return betAmount * 2;
    }

    return 0;
  }
}
