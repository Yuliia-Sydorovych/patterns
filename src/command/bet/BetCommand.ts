import { UIManager } from "../../ui/UIManager";
import { ICommand } from "../interface/ICommand";
import { BetManager } from "./BetManager";

export class BetCommand implements ICommand
{
  private betManager: BetManager;
  private uiManager: UIManager;

  constructor(betManager: BetManager, uiManager: UIManager)
  {
    this.betManager = betManager;
    this.uiManager = uiManager;
  }

  public execute(value: number): void
  {
    const result = this.betManager.checkBet(value) ? `Bet of $${value} placed successfully.` : "Insufficient funds to place the bet.";
    this.uiManager.updateGameResult(result);
  }
}
