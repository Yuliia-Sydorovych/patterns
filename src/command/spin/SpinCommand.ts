import { SlotMachineManager } from "../../reels/SlotMachineManager";
import { UIManager } from "../../ui/UIManager";
import { BetManager } from "../bet/BetManager";
import { BaseCommand } from "../BaseCommand";

export class SpinCommand extends BaseCommand
{
  private slotMachineManager: SlotMachineManager;

  constructor(betManager: BetManager, slotMachineManager: SlotMachineManager, uiManager: UIManager)
  {
    super(betManager, uiManager);

    this.slotMachineManager = slotMachineManager;
  }

  public execute(betAmount: number): void
  {
    if (this.betManager.placeBet(betAmount))
    {
      this.uiManager.updateOnSpinStart();

      this.slotMachineManager.spin(() =>
      {
        const payout = this.slotMachineManager.calculatePayout(betAmount);
        this.betManager.updateBalance(payout);
        this.uiManager.updateOnSpinEnd(payout);
      });
    }
    else
    {
      this.uiManager.showInsufficientFunds();
    }
  }
}
