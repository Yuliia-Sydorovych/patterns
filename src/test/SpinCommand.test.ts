import { BetManager } from "../command/bet/BetManager";
import { SpinCommand } from "../command/spin/SpinCommand";
import { SlotMachineManager } from "../reels/SlotMachineManager";
import { UIManager } from "../ui/UIManager";

describe('SpinCommand', () =>
{
  let betManager: jest.Mocked<BetManager>;
  let slotMachineManager: jest.Mocked<SlotMachineManager>;
  let uiManager: jest.Mocked<UIManager>;
  let spinCommand: SpinCommand;

  beforeEach(() =>
  {
    betManager =
    {
      placeBet: jest.fn(),
      updateBalance: jest.fn(),
    } as unknown as jest.Mocked<BetManager>;

    slotMachineManager =
    {
      spin: jest.fn(),
      calculatePayout: jest.fn(),
    } as unknown as jest.Mocked<SlotMachineManager>;

    uiManager =
    {
      updateOnSpinStart: jest.fn(),
      updateOnSpinEnd: jest.fn(),
      showInsufficientFunds: jest.fn(),
    } as unknown as jest.Mocked<UIManager>;

    spinCommand = new SpinCommand(betManager, slotMachineManager, uiManager);
  });

  it('should spin and update balance when bet is placed successfully', () =>
  {
    const betAmount = 100;
    const payout = 250;
    betManager.placeBet.mockReturnValue(true);
    slotMachineManager.calculatePayout.mockReturnValue(payout);

    spinCommand.execute(betAmount);

    expect(betManager.placeBet).toHaveBeenCalledWith(betAmount);
    expect(uiManager.updateOnSpinStart).toHaveBeenCalled();

    expect(slotMachineManager.spin).toHaveBeenCalled();
    const spinCallback = slotMachineManager.spin.mock.calls[0][0];
    spinCallback(); 

    expect(slotMachineManager.calculatePayout).toHaveBeenCalledWith(betAmount);
    expect(betManager.updateBalance).toHaveBeenCalledWith(payout);
    expect(uiManager.updateOnSpinEnd).toHaveBeenCalledWith(payout);

    expect(uiManager.updateOnSpinEnd).toMatchSnapshot();
  });

  it('should show insufficient funds if bet cannot be placed', () =>
  {
    const betAmount = 100;
    betManager.placeBet.mockReturnValue(false);

    spinCommand.execute(betAmount);

    expect(betManager.placeBet).toHaveBeenCalledWith(betAmount);
    expect(uiManager.showInsufficientFunds).toHaveBeenCalled();
    expect(uiManager.updateOnSpinStart).not.toHaveBeenCalled();
    expect(slotMachineManager.spin).not.toHaveBeenCalled();
    expect(uiManager.updateOnSpinEnd).not.toHaveBeenCalled();

    expect(uiManager.showInsufficientFunds).toMatchSnapshot();
  });
});
