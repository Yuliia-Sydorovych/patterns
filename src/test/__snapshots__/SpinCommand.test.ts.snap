// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SpinCommand should show insufficient funds if bet cannot be placed 1`] = `
[MockFunction] {
  "calls": [
    [],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`SpinCommand should spin and update balance when bet is placed successfully 1`] = `
[MockFunction] {
  "calls": [
    [
      250,
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
